PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
 Code related to processing of import hooks.
 """
 
-import glob
+import glob, weakref
 import os.path
 
 from .. import log as logging
-from .utils import format_binaries_and_datas
-from ..compat import expand_path
-from ..compat import importlib_load_source
+from ..compat import (
+    expand_path, importlib_load_source, FileNotFoundError, UserDict)
 from .imphookapi import PostGraphAPI
+from .utils import format_binaries_and_datas
 
 logger = logging.getLogger(__name__)
 
 
+class ModuleHookCache(UserDict):
+    """
+    Cache of lazily loadable hook script objects.
+
+    This cache is implemented as a `dict` subclass mapping from the
+    fully-qualified names of all modules with at least one hook script to lists
+    of `ModuleHook` instances encapsulating these scripts. As a `dict` subclass,
+    all cached module names and hook scripts are accessible via standard
+    dictionary operations.
+
+    Attributes
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    _hook_module_name_prefix : str
+        String prefixing the names of all in-memory modules lazily loaded from
+        cached hook scripts. See also the `hook_module_name_prefix` parameter
+        passed to the `ModuleHook.__init__()` method.
+    """
+
+    _cache_id_next = 0
+    """
+    0-based identifier unique to the next `ModuleHookCache` to be instantiated.
+
+    This identifier is incremented on each instantiation of a new
+    `ModuleHookCache` to isolate in-memory modules of lazily loaded hook script
+    code objects in that cache to the same cache-specific namespace, preventing
+    edge-case collisions with existing in-memory modules in other caches.
+    """
+
+    def __init__(self, module_graph):
+        """
+        Initialize this cache.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        """
+
+        # Curiously, calling "super(UserDict, self).__init__()" here instead
+        # subsequently results in exceptions resembling:
+        #     AttributeError: 'ModuleHookCache' object has no attribute 'data'
+        UserDict.__init__(self)
+
+        # To avoid circular references and hence increased memory consumption,
+        # a weak rather than strong reference is stored to the passed graph.
+        # Since this graph is guaranteed to live longer than this cache, this is
+        # guaranteed to be safe.
+        self.module_graph = weakref.proxy(module_graph)
+
+        # String unique to this cache prefixing the names of all in-memory
+        # modules lazily loaded from cached hook scripts.
+        self._hook_module_name_prefix = 'PyInstaller_hooks_{}_'.format(
```

Comment:
Two underscores! Now you're playing with power. It shall be so, for it is a healthy idea.